README.md
[![Java CI with Maven](https://github.com/Maritims/node-maven-plugin/actions/workflows/maven.yml/badge.svg)](https://github.com/Maritims/node-maven-plugin/actions/workflows/maven.yml)
[![Maven Central](https://maven-badges.herokuapp.com/maven-central/io.github.maritims/node-maven-plugin/badge.svg)](https://maven-badges.herokuapp.com/maven-central/io.github.maritims/node-maven-plugin)

# node-maven-plugin
Maven plugin for running node commands. This plugin will download and extract [Node.js](https://nodejs.dev/) in the directory you specify to allow you to run Node.js commands such as `npm`. See [node-maven-plugin-poc-webapp](https://github.com/Maritims/node-maven-plugin-poc-webapp) for sample usage.  

## Configuration
Add the plugin groupId to the `<pluginGroups />` section of settings.xml. This tells Maven where to look when resolving `mvn` plugin commands and lets you type `mvn node:npm` rather than the long form.
See [Maven - Guide to Developing Java Plugins](https://maven.apache.org/guides/plugin/guide-java-plugin-development.html) for more details.
```xml
<pluginGroups>
    <pluginGroup>io.github.maritims</pluginGroup>
</pluginGroups>
```

Add the plugin configuration to the pom.xml in your Maven project:
```xml
<build>
    <plugins>
        <plugin>
            <groupId>io.github.maritims</groupId>
            <artifactId>node-maven-plugin</artifactId>
            <version>1.1.0</version>
            <configuration>
                <install>true</install>
                <script>build</script>
                <sourceCodePath>svelte</sourceCodePath>
            </configuration>
            <executions>
                <execution>
                    <goals>
                        <goal>npm</goal>
                    </goals>
                </execution>
            </executions>
        </plugin>
    </plugins>
</build>
```

## Options
| Option                  | Type    | Value                               | Default value | Explanation                                                                                                                 |
|-------------------------|---------|-------------------------------------|---------------|-----------------------------------------------------------------------------------------------------------------------------|
| nodePath                | string  | node                                | node          | Path to directory within project base directory where node should be downloaded and installed.                              |
| major                   | int     | 16                                  | 16            | Major version of node.                                                                                                      |
| minor                   | int     | 14                                  | 14            | Minor version of node.                                                                                                      |
| patch                   | int     | 2                                   | 2             | Patch version of node.                                                                                                      |
| sourceCodeDirectoryName | string  | svelte, node, vue, react            | N/A           | Path to directory within the src directory containing package.json. The directory should live alongside the java directory. | 
| install                 | boolean | true, false                         | N/A           | Indicates whether packages should be installed.                                                                             |
| script                  | string  | build                               | N/A           | Indicates which script to run. The script must be present in package.json.                                                  |
| environmentVariables    | string  | FOO_BAR=foo BAR_BAZ=bar BAZ_BAR=baz | N/A           | Environment variables for the npm build process.                                                                            |

## Install npm packages
`mvn node:npm -DsourceCodeDirectoryName=<sourceCodeDirectoryName> -Dinstall`

## Run npm script
`mvn node:npm -DsourceCodeDirectoryName=<sourceCodeDirectoryName> -Dscript <script>`

## Define output directory for npm build
How you define the output directory for your npm build depends on the bundling software you decide to use. You typically want the npm build to put the file(s) somewhere in src/main/webapp, or a sub-directory.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.0] - 2022-10-09
### Added
- `NpmMojo` now accepts the property `environmentVariables` in the format `FOO_BAR=helloworld BAR_BAZ=loremipsum` and so on, just like you would prefix an `npm` command on the command line.

## [1.0.3] - 2022-24-30
### Added
- Added node bin directory to ProcessBuilder path environment variable.
- Created Dockerfile to test project without pollution from host system.
- `NodeWrapper.extract` now takes a `boolean` argument `verbose` which currently only dictates whether to verbosely log archive extraction.
- `PackageJson.get` now handles exceptions related to opening the file `package.json` and deserialization attempts.

### Changed
- Extraction is now skipped if the target directory already exists.

### Removed
- Removed unnecessary `IOException` from `NpmWrapper.run` method signature.

## [1.0.2] - 2022-04-25
### Added
- Added javadoc, source and gpg plugins

### Changed
- Group id was changed from com.github.maritims to io.github.maritims
- CI now uses `mvn verify` instead of `mvn package`

## [1.0.1] - 2022-04-24
### Added
- Npm mojo.
- Npm wrapper.
- Node installer.
- Initial commit.

### Removed
- Yarn mojo.

[Unreleased]: https://github.com/Maritims/node-maven-plugin/compare/node-maven-plugin-1.0.3...HEAD
[1.1.0]: https://github.com/Maritims/node-maven-plugin/compare/node-maven-plugin-1.0.3...node-maven-plugin-1.1.0
[1.0.3]: https://github.com/Maritims/node-maven-plugin/compare/node-maven-plugin-1.0.2...node-maven-plugin-1.0.3
[1.0.2]: https://github.com/Maritims/node-maven-plugin/compare/node-maven-plugin-1.0.1...node-maven-plugin-1.0.2
[1.0.1]: https://github.com/Maritims/node-maven-plugin/compare/node-maven-plugin-1.0.0...node-maven-plugin-1.0.1
[1.0.0]: https://github.com/Maritims/node-maven-plugin/releases/tag/node-maven-plugin-1.0.0

src/main/java/io/github/maritims/NpmMojo.java
package io.github.maritims;

import io.github.maritims.node.NodeConfiguration;
import org.apache.maven.plugin.AbstractMojo;
import org.apache.maven.plugin.MojoExecutionException;
import org.apache.maven.plugin.MojoFailureException;
import org.apache.maven.plugins.annotations.LifecyclePhase;
import org.apache.maven.plugins.annotations.Mojo;
import org.apache.maven.plugins.annotations.Parameter;
import org.apache.maven.project.MavenProject;

import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Mojo(name = "npm", defaultPhase = LifecyclePhase.INSTALL)
public class NpmMojo extends AbstractMojo {
    @Parameter(defaultValue = "${project}", required = true, readonly = true)
    MavenProject project;

    @Parameter(property = "nodePath", required = true, defaultValue = "node")
    String nodePath;

    @Parameter(property = "major", required = true, defaultValue = "16")
    int major;

    @Parameter(property = "minor", required = true, defaultValue = "14")
    int minor;

    @Parameter(property = "patch", required = true, defaultValue = "2")
    int patch;

    @Parameter(property = "sourceCodeDirectoryName", required = true)
    String sourceCodeDirectoryName;

    @Parameter(property = "install", required = true, defaultValue = "false")
    boolean install;

    @Parameter(property = "script", required = true)
    String script;

    @Parameter(property = "environmentVariables")
    String environmentVariables;

    public void execute() throws MojoExecutionException, MojoFailureException {
        NpmWrapper npm = new NpmWrapper(
                new NodeConfiguration(Paths.get(nodePath), Paths.get(nodePath), major, minor, patch),
                Paths.get(project.getBasedir().getAbsolutePath(), "src", "main", sourceCodeDirectoryName).toString()
        );

        if(install) npm.install();
        npm.run(script, environmentVariables);
    }
}
